sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -4819,148 +4825,181 @@ case class ArrayInsert(srcArrayExpr: Expression, posExpr: Expression, itemExpr:
-         |    $adjustedAllocIdx = $i;
-         |    if ($i >= $itemInsertionIndex) {
-         |      $adjustedAllocIdx = $adjustedAllocIdx + 1;
-         |    }
-         |    $assignment
-         |  }
-         |  ${CodeGenerator.setArrayElement(
-              values, elementType, itemInsertionIndex, item, Some(insertedItemIsNull))}
-         |
-         |  for (int $j = $arr.numElements(); $j < $resLength - 1; $j ++) {
-         |    $values.setNullAt($j);
-         |  }
-         |
-         |  ${ev.value} = $values;
-         |}
-      """.stripMargin
+      if (canBeOptimizedForPrepend) {
+        val zero = "0"
+        s"""
+           |int $resLength = $arr.numElements() + 1;
+           |int $adjustedAllocIdx = 0;
+           |if ($resLength > ${ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH}) {
+           |  throw QueryExecutionErrors.createArrayWithElementsExceedLimitError($resLength);
+           |}
+           |boolean $insertedItemIsNull = ${itemExpr.isNull};
+           |
+           |$allocation
+           |for (int $i = 0; $i < $arr.numElements(); $i ++) {
+           |  $adjustedAllocIdx = $i + 1;
+           |  $assignment
+           |}
+           |${CodeGenerator.setArrayElement(
+              values, elementType, zero, item, Some(insertedItemIsNull))}
+           |
+           |${ev.value} = $values;
+           |""".stripMargin
+      } else {
+        val pos = posExpr.value
+        val itemInsertionIndex = ctx.freshName("itemInsertionIndex")
+        val j = ctx.freshName("j")
+        val errorContext = getContextOrNullCode(ctx)
+        s"""
+           |int $itemInsertionIndex = 0;
+           |int $resLength = 0;
+           |int $adjustedAllocIdx = 0;
+           |boolean $insertedItemIsNull = ${itemExpr.isNull};
+           |
+           |if ($pos == 0) {
+           |  throw QueryExecutionErrors.invalidIndexOfZeroError($errorContext);
+           |}
+           |
+           |if ($pos < 0 && (java.lang.Math.abs($pos) > $arr.numElements())) {
+           |
+           |  $resLength = java.lang.Math.abs($pos) + 1;
+           |  if ($resLength > ${ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH}) {
+           |    throw QueryExecutionErrors.createArrayWithElementsExceedLimitError($resLength);
+           |  }
+           |
+           |  $allocation
+           |  for (int $i = 0; $i < $arr.numElements(); $i ++) {
+           |    $adjustedAllocIdx = $i + 1 + java.lang.Math.abs($pos + $arr.numElements());
+           |    $assignment
+           |  }
+           |  ${CodeGenerator.setArrayElement(
+                values, elementType, itemInsertionIndex, item, Some(insertedItemIsNull))}
+           |
+           |  for (int $j = $pos + $arr.numElements(); $j < 0; $j ++) {
+           |    $values.setNullAt($j + 1 + java.lang.Math.abs($pos + $arr.numElements()));
+           |  }
+           |
+           |  ${ev.value} = $values;
+           |} else {
+           |
```

Comment:
can you take a close look at the generated code and see how we can simplify it if the position is a constant value? By just 10 second look, I found that if the position is constant and positive, the code can be simplified to
```
final int $resLength = java.lang.Math.max($arr.numElements() + 1, $pos);
...
```

Let's not be so specific that only target the case when `position` is `1`. We can generate code when position is constant and positive, which is a very common case.

Reply:
@cloud-fan https://github.com/apache/spark/pull/40833 used to replace this one.